### PR TITLE
move exec.reshape to backend

### PR DESCRIPTION
**Pull request by ZiyueHuang**
## Description ##
move exec.reshape to backend

cc @reminisce @eric-haibin-lin @piiswrong 

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/c_api/c_api_executor.cc [hunk 1]
```diff
@@ -510,6 +513,132 @@ int MXExecutorSimpleBind(SymbolHandle symbol_handle,
   API_END();
 }
 
+int MXExecutorReshapeEX(SymbolHandle symbol_handle,
```
Comment (eric-haibin-lin) [inline src/c_api/c_api_executor.cc]: nit: MXExecutorReshapeEX -> MXExecutorReshapeEx

File: include/mxnet/c_api.h [hunk 2]
```diff
@@ -1647,6 +1647,31 @@ MXNET_DLL int MXExecutorSimpleBind(SymbolHandle symbol_handle,
                                    NDArrayHandle** aux_states,
                                    ExecutorHandle shared_exec_handle,
                                    ExecutorHandle* out);
+/*
+ *! \brief Return a new executor with the same symbol and shared memory,
+ *         but different input/output shapes.
+ */
+MXNET_DLL int MXExecutorReshapeEx(SymbolHandle symbol_handle,
```
Comment (piiswrong) [inline include/mxnet/c_api.h]: why Ex? is there a MXExecutorReshape already?

File: src/c_api/c_api_executor.cc [hunk 3]
```diff
@@ -510,6 +513,135 @@ int MXExecutorSimpleBind(SymbolHandle symbol_handle,
...
+  CHECK_EQ(g.GetAttr<size_t>("shape_num_unknown_nodes"), 0U);
+  const std::vector<TShape>& shape_vec = g.GetAttr<std::vector<TShape>>("shape");
+  NDArrayHandle *arg = in_args;
+  NDArrayHandle *grad = arg_grad_store;
+  NDArrayHandle *aux = aux_states;
+
+  for (uint32_t nid : idx.input_nodes()) {
+    std::string name = idx[nid].source->attrs.name;
+    const TShape& new_shape = shape_vec[idx.entry_id(nid, 0)];
+    if (idx.mutable_input_nodes().count(nid) == 0) {
```
Comment (reminisce) [inline src/c_api/c_api_executor.cc]: Can you consolidate the implementation for args and aux? There is a big chunk of duplicate code.

File: src/c_api/c_api_executor.cc [hunk 4]
```diff
@@ -510,6 +513,135 @@ int MXExecutorSimpleBind(SymbolHandle symbol_handle,
...
+      NDArray* arr = static_cast<NDArray*>(*arg);
+      NDArray* darr = static_cast<NDArray*>(*grad);
+      if (new_shape == arr->shape()) {
```
Comment (reminisce) [inline src/c_api/c_api_executor.cc]: Please simplify this. There is no need to have a `do nothing` block.

File: src/c_api/c_api_executor.cc [hunk 5]
```diff
@@ -510,6 +513,135 @@ int MXExecutorSimpleBind(SymbolHandle symbol_handle,
...
+        // do nothing
+      } else if (partial_shaping || kwargs.count(name)) {
+        if (new_shape.Size() > arr->shape().Size()) {
+          CHECK(allow_up_sizing) << "New shape of arg:" << name
+              << " larger than original. "
+              << "First making a big executor and then down sizing it "
+              << "is more efficient than the reverse."
+              << "If you really want to up size, set allow_up_sizing=True "
+              << "to enable allocation of new arrays.";
+          NDArray* empty_arr = new NDArray(new_shape, arr->ctx(), false, arr->dtype());
```
Comment (reminisce) [inline src/c_api/c_api_executor.cc]: This looks like a memory leak. Please use `arr.ReshapeAndAlloc(new_shape)`.

File: src/c_api/c_api_executor.cc [hunk 6]
```diff
@@ -510,6 +513,135 @@ int MXExecutorSimpleBind(SymbolHandle symbol_handle,
...
+        // do nothing
+      } else if (partial_shaping) {
+        if (new_shape.Size() > arr->shape().Size()) {
+          CHECK(allow_up_sizing) << "New shape of arg:" << name
+              << " larger than original. "
+              << "First making a big executor and then down sizing it "
+              << "is more efficient than the reverse."
+              << "If you really want to up size, set allow_up_sizing=True "
+              << "to enable allocation of new arrays.";
+          NDArray* empty_arr = new NDArray(new_shape, arr->ctx(), false, arr->dtype());
```
Comment (reminisce) [inline src/c_api/c_api_executor.cc]: Same here. ReshapeAndAlloc.

**Comment by ZiyueHuang** [comment]
Thanks for your comments.
For the unit tests part, I found the existing tests cover both up sizing and down sizing, https://github.com/apache/incubator-mxnet/blob/master/tests/python/unittest/test_module.py#L703, https://github.com/apache/incubator-mxnet/blob/master/tests/python/unittest/test_module.py#L205, https://github.com/apache/incubator-mxnet/blob/master/tests/python/unittest/test_executor.py#L144.
Will address other parts tomorrow.

File: python/mxnet/executor.py [hunk 7]
```diff
@@ -399,62 +399,96 @@ def reshape(self, partial_shaping=False, allow_up_sizing=False, **kwargs):
...
+        provided_arg_shape_names = []  # provided argument names
+        for k, v in kwargs.items():
+            # if k not in listed_arguments and k not in listed_aux_states:
+            #   raise ValueError('arg name %s is not valid', k)
+            if isinstance(v, tuple):
+                provided_arg_shape_names.append(k)
+                provided_arg_shape_data.extend(v)
+                provided_arg_shape_idx.append(len(provided_arg_shape_data))
+
+        args_handle, args = self._symbol._get_ndarray_inputs(
```
Comment (reminisce) [inline python/mxnet/executor.py]: I think you can get all the arg/aux/grad arrays from the executor in backend? Is it necessary to prepare them here and pass to the backend again?

**Comment by ZiyueHuang** [comment]
Could you please review again? I realized that `Reshape` is more appropriate to be a member function of `Executor` so that it can access the internal arguments of the executor. Also I think `ReshapeAndAlloc` should not be used, since the original executor should work fine after reshape.

**Comment by ZiyueHuang** [comment]
unittest for checking
>whether the reshaped executor and the original executor share NDArrays depending on the new shape sizes

is added in `test_executor.py` @reminisce

File: src/executor/graph_executor.cc [hunk 8]
```diff
@@ -1043,6 +1043,108 @@ void GraphExecutor::Init(nnvm::Symbol symbol,
...
+                                 std::vector<NDArray>* arg_grads,
+                                 std::vector<NDArray>* aux_states) {
+  nnvm::Graph g;
+  g.outputs = std::vector<nnvm::NodeEntry>(graph_.outputs.begin(),
+    graph_.outputs.begin() + num_forward_outputs_);
+  nnvm::Symbol symbol;
+  symbol.outputs = g.outputs;
+  const nnvm::IndexedGraph& idx = g.indexed_graph();
+  nnvm::ShapeVector arg_shapes(idx.input_nodes().size(), TShape());
+  for (size_t i = 0; i < num_forward_inputs_; ++i) {
```
[the remark this answers is not in the document]
Comment (ZiyueHuang) [reply src/executor/graph_executor.cc]: changed

**Comment by ZiyueHuang** [comment]
ping @reminisce @piiswrong

**Comment by szha** [comment]
ping @piiswrong

**Comment by piiswrong** [comment]
LGTM. Any concerns? @reminisce @eric-haibin-lin

**Pull request merged**